in-progress/8401-proof-timeliness/design.md

```diff
@@ -0,0 +1,268 @@
+        uint256 finalBlockNumber;
+        Fee[] fees;
+    }
+
+    // State variables
+    State public state;
+    mapping(uint256 => ProposalLog) public proposalLogs;
+    ProofClaim public proofClaim;
+    IERC20 public testToken;
+    IFeeJuicePortal public feeJuicePortal;
+
+    // Constants
+    uint256 public constant CLAIM_DURATION = 13;
+    uint256 public constant EPOCH_DURATION = 32;
+    uint256 public constant PROOF_COMMITMENT_BOND_AMOUNT = 100;
+
+    // Events
+    event ProofRightClaimed(address indexed claimer, uint256 indexed epoch);
+    event PendingChainPruned(uint256 blockNumber, uint256 slotNumber);
+    event ProofSubmitted(uint256 indexed epoch, uint256 finalBlockNumber);
+
+    // Errors
+    error Rollup__NonProposerCannotClaimProofRight();
+    error Rollup__ProofRightAlreadyClaimed();
+    error Rollup__NotInClaimPhase();
+    error Rollup__NoProposalLog();
+    error Rollup__NotPreviousEpoch();
+    error Rollup__NotLastBlockInEpoch();
+    error Rollup__ProofProductionPhaseEnded();
+    error Rollup__UnauthorizedProofSubmitter();
+
+    constructor(address _testToken, address _feeJuicePortal) {
+        testToken = IERC20(_testToken);
+        feeJuicePortal = IFeeJuicePortal(_feeJuicePortal);
+    }
+
+    function claimProofRight(address _bondProvider) external {
+        uint256 currentSlot = getCurrentSlot();
+        address currentProposer = getCurrentProposer();
+
+        if (currentProposer != address(0) && currentProposer != msg.sender) {
+            revert Rollup__NonProposerCannotClaimProofRight();
+        }
+
+        if (proofClaim.rewardRecipient != address(0)) {
+            revert Rollup__ProofRightAlreadyClaimed();
+        }
+
+        if (currentSlot % EPOCH_DURATION >= CLAIM_DURATION) {
+            revert Rollup__NotInClaimPhase();
+        }
+
+        // Bond provider must have approved the contract to transfer the bond amount
+        testToken.transferFrom(_bondProvider, address(this), PROOF_COMMITMENT_BOND_AMOUNT);
+
+        proofClaim = ProofClaim({
+            rewardRecipient: msg.sender,
+            bondProvider: _bondProvider,
+            claimedEpoch: getCurrentEpoch() - 1
+        });
+
+        emit ProofRightClaimed(msg.sender, proofClaim.claimedEpoch, currentSlot);
+    }
+
+    function _prune() internal {
+        uint256 currentSlot = getCurrentSlot();
+        uint256 currentEpoch = getCurrentEpoch();
+        uint256 currentEpochStartSlot = currentEpoch * EPOCH_DURATION;
+
+        if (proofClaim.rewardRecipient == address(0) &&
+            currentSlot >= currentEpochStartSlot + CLAIM_DURATION) {
+            state.pendingTip = state.provenTip;
+            emit PendingChainPruned(state.pendingTip.blockNumber, state.pendingTip.slotNumber);
+        } else if (proofClaim.rewardRecipient != address(0) &&
+                   currentSlot >= currentEpochStartSlot + EPOCH_DURATION) {
+            state.pendingTip = state.provenTip;
+            _slashBond();
```

Comment:
I don't think you actually need a separate function to "slash" bond. As long as it "gets lost" by it being pruned. Think accounting can be done somewhat smart such that we don't have to actually update storage here. Which also means that the two cases become just one which seems to make it slightly simpler to read.